### ISSUE TITLE
Inference demo

### ISSUE THREAD

**Issue by dzyjjpy** [open-mmlab/mmrazor#165]
Hi,

Is there a inference demo in the code base? Try to init model follow test/test_apis/test_inference.py.

`result = inference_detector(model, img_path)
show_result_pyplot(model, img_path, result, wait_time=1, score_thr=0.7, palette=(72, 101, 241))
`

The terminal shows model and checkpoints loaded state dict do not match exactly.  
The bbox painted on image is not correct, and every time is different.
However, I'm sure the mAP tested by test.py is correct.

**Comment by pppppM** [comment]
Thanks for your attention.
Can you show the warning message during load state dict?

**Comment by dzyjjpy** [comment]
@pppppM Hi, as below
`
load checkpoint from local path: work_dirs/detnas_subnet_retinanet_sfv2_fpn_1x_coco_1121_n8h8f8_16_32_64_96_hgrhead_3anchors_conv8/epoch_15.82.3.pth
The model and loaded state dict do not match exactly
unexpected key in source state_dict: architecture.model.backbone.conv1.conv.weight, architecture.model.backbone.conv1.bn.weight, architecture.model.backbone.conv1.bn.bias, architecture.model.backbone.conv1.bn.running_mean, architecture.model.backbone.conv1.bn.running_var, architecture.model.backbone.conv1.bn.num_batches_tracked, architecture.model.backbone.layers.0.0.choices.shuffle_3x3.branch1.0.conv.weight, architecture.model.backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.weight, architecture.model.backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.bias, architecture.model.backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.running_mean, architecture.model.backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.running_var, architecture.model.backbone.layers.0.0.choices.shuffle_3x3.branch1.0 ........
anaconda3/envs/mmrazor/lib/python3.8/site-packages/mmdet/datasets/utils.py:66: UserWarning: "ImageToTensor" pipeline is replaced by "DefaultFormatBundle" for batch inference. It is recommended to manually replace it in the test data pipeline in your config file.`

eval mAP shows:
`2022-05-24 10:15:36,442 - mmdet - INFO -
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.419
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=1000 ] = 0.830
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=1000 ] = 0.348
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.015
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.408
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.453
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.510
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=300 ] = 0.510
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=1000 ] = 0.510
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=1000 ] = 0.078
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=1000 ] = 0.521
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=1000 ] = 0.513
`

use --show-dir to save, result has no problem. FYI 
Thanks :)

**Comment by pppppM** [comment]
Are there only unexpected keys, not missing keys?

**Comment by dzyjjpy** [comment]
> Are there only unexpected keys, not missing keys?

I check the whole information, there is missing keys warning:

`missing keys in source state_dict: backbone.conv1.conv.weight, backbone.conv1.bn.weight, backbone.conv1.bn.bias, backbone.conv1.bn.running_mean, backbone.conv1.bn.running_var, backbone.layers.0.0.choices.shuffle_3x3.branch1.0.conv.weight, backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.weight, backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.bias, backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.running_mean, backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.running_var, backbone.layers.0.0.choices.shuffle_3x3.branch1.1.conv.weight, backbone.layers.0.0.choices.shuffle_3x3.branch1.1.bn.weight, backbone.layers.0.0.choices.shuffle_3x3.branch1.1.bn.bias, backbone.layers.0.0.choices.shuffle_3x3.branch1.1.bn.running_mean, backbone.layers.0.0.choices.shuffle_3x3.branch1.1.bn.running_var, backbone.layers.0.0.choices.shuffle_3x3.branch2.0.conv.weight, backbone.layers.0.0.choices.shuffle_3x3.branch2.0.bn.weight
......
`

**Comment by dzyjjpy** [comment]
> > Are there only unexpected keys, not missing keys?
> 
> I check the whole information, there is missing keys warning:
> 
> `missing keys in source state_dict: backbone.conv1.conv.weight, backbone.conv1.bn.weight, backbone.conv1.bn.bias, backbone.conv1.bn.running_mean, backbone.conv1.bn.running_var, backbone.layers.0.0.choices.shuffle_3x3.branch1.0.conv.weight, backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.weight, backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.bias, backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.running_mean, backbone.layers.0.0.choices.shuffle_3x3.branch1.0.bn.running_var, backbone.layers.0.0.choices.shuffle_3x3.branch1.1.conv.weight, backbone.layers.0.0.choices.shuffle_3x3.branch1.1.bn.weight, backbone.layers.0.0.choices.shuffle_3x3.branch1.1.bn.bias, backbone.layers.0.0.choices.shuffle_3x3.branch1.1.bn.running_mean, backbone.layers.0.0.choices.shuffle_3x3.branch1.1.bn.running_var, backbone.layers.0.0.choices.shuffle_3x3.branch2.0.conv.weight, backbone.layers.0.0.choices.shuffle_3x3.branch2.0.bn.weight ...... `

@pppppM Any advice?

**Comment by pppppM** [comment]
Is the ckpt you loaded trained with mmrazor? 
Normally, all keys should be prefixed with `architecture.model`

**Comment by dzyjjpy** [comment]
@pppppM FYI :)

**Comment by pppppM** [comment]
The checkpoint should be loaded in `init_mmdet_model`.

https://github.com/open-mmlab/mmrazor/blob/2dad24044d7f1dad88f20221f8fc071dd40fdd4f/mmrazor/apis/mmdet/inference.py#L13-L16
This function's motivation is to export a detector that is consistent with mmdet, so the returned model will be inconsistent with razor's ckpt. 
We load ckpt inside `init_mmdet_model`
https://github.com/open-mmlab/mmrazor/blob/2dad24044d7f1dad88f20221f8fc071dd40fdd4f/mmrazor/apis/mmdet/inference.py#L45-L56

**Comment by pppppM** [comment]
Sorry, I didn't notice the last question because the issue was closed.
The algorithm in razor is more like a wrapper around the detector. In order to better reuse the various tools in mmdet, the algorithm needs to be converted to a detector.
Since a detector is trained with the algorithm in razor, the module name will have the prefix `algorithm.model` than the original detector. 
So there needs to load ckpt before converting the algorithm to the detector.
